src/compiler/checker.ts

```diff
@@ -573,7 +573,16 @@ namespace ts {
                         declarationNameToString(propertyName), typeof nameArg === "string" ? nameArg : declarationNameToString(nameArg));
                     return undefined;
                 }
-                if (result.flags & SymbolFlags.BlockScopedVariable) {
+
+                // Only check for block-scoped variable if we are looking for the name with variable meaning
+                // For example,
+                //      declare module foo {
+                //          interface bar {}
+                //      }
+                // let foo/*1*/: foo/*2*/.bar;
+                // The foo at /*1*/ and /*2*/ will share same symbol with two meaning block-scope variable and namespace module.
+                // However, only when we try to resolve name in /*1*/ which is used in variable position, we want to check for block- scoped

```

Comment:
@DanielRosenwasser I think your suggested assert would actually fail when you find an alias that references the meaning you're looking for. Like this:
```ts
module X.Y { ; } // Instantiated
import a = X.Y;
var b = a; // call resolveName for 'a'
```
In this case, you'd be looking for a value, but you would get an alias (to a value).

I think the comment is good.